### PR TITLE
HIVE-24654: Table level replication support for Atlas metadata

### PR DESCRIPTION
**Pull request by pkumarsinha**

### CODE REVIEW

File: ql/src/test/org/apache/hadoop/hive/ql/exec/repl/TestAtlasDumpTask.java [hunk 1]
```diff
@@ -242,6 +251,25 @@ public void testAtlasClientTimeouts() throws Exception {
             AtlasRestClientBuilder.ATLAS_PROPERTY_READ_TIMEOUT_IN_MS));
   }
 
+  @Test
+  public void testCreateExportRequest() throws Exception {
```
Comment (aasha) [inline ql/src/test/org/apache/hadoop/hive/ql/exec/repl/TestAtlasDumpTask.java:255]: what happens if policy is modified to remove the table expression or modify the table expression? Does it fallback to db level replication?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/AtlasLoadTask.java [hunk 2]
```diff
@@ -121,7 +121,7 @@ AtlasReplInfo createAtlasReplInfo() throws SemanticException, MalformedURLExcept
     String srcCluster = ReplUtils.getNonEmpty(HiveConf.ConfVars.REPL_SOURCE_CLUSTER_NAME.varname, conf, errorFormat);
     String tgtCluster = ReplUtils.getNonEmpty(HiveConf.ConfVars.REPL_TARGET_CLUSTER_NAME.varname, conf, errorFormat);
     AtlasReplInfo atlasReplInfo = new AtlasReplInfo(endpoint, work.getSrcDB(), work.getTgtDB(),
-            srcCluster, tgtCluster, work.getStagingDir(), conf);
+            srcCluster, tgtCluster, work.getStagingDir(), null, conf);
```
[the remark this answers is not in the document]
Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/AtlasLoadTask.java:124]: Done

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/atlas/AtlasRequestBuilder.java [hunk 3]
```diff
@@ -105,6 +126,50 @@ private String getQualifiedName(String clusterName, String srcDb) {
...
+      qualifiedNames.add(getQualifiedName(clusterName, srcDb, tableName));
+    }
+    return qualifiedNames;
+  }
+
+  private static List<String> getFileAsList(Path listOfTablesFile, HiveConf conf) throws SemanticException {
+    List<String> list = new ArrayList<>();
+    InputStream is = null;
+    try {
+      FileSystem fs = FileSystem.get(listOfTablesFile.toUri(), conf);
```
[the remark this answers is not in the document]
Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/atlas/AtlasRequestBuilder.java:152]: No, doesn't look like. The FS creation looks proper so should work in those cases.

File: ql/src/test/org/apache/hadoop/hive/ql/exec/repl/TestAtlasDumpTask.java (hunk 1 above)
Comment (pkumarsinha) [reply ql/src/test/org/apache/hadoop/hive/ql/exec/repl/TestAtlasDumpTask.java:255]: Hive replication as such will not allow falling back from table level to db level replication. For a expression modification case, Atlas server should handle the case and if it does not. we will have to get that fixed by Atlas team. For now, there is no known case on Atlas server side.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/atlas/AtlasRequestBuilder.java [hunk 4]
```diff
@@ -105,6 +126,50 @@ private String getQualifiedName(String clusterName, String srcDb) {
...
+      LOG.info("Empty file encountered: {}", listOfTablesFile);
+      return qualifiedNames;
+    }
+    for (String tableName : tableNames) {
+      qualifiedNames.add(getQualifiedName(clusterName, srcDb, tableName));
+    }
+    return qualifiedNames;
+  }
+
+  public List<String> getFileAsList(Path listOfTablesFile, HiveConf conf) throws SemanticException {
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/atlas/AtlasRequestBuilder.java:148]: Does this need retry?

Comment (pkumarsinha) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/atlas/AtlasRequestBuilder.java:149]: Done

**Review by aasha — APPROVED** [review]

**Pull request merged**